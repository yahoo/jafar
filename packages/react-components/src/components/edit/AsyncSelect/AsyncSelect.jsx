/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import SelectInternal from 'react-select';
import PropTypes from 'prop-types';
import { customStyles } from '../Select/Select';

/**
 * Represent a selection of a single item from async items list
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/AsyncSelect/AsyncSelect.jsx">AsyncSelect</a>
 from '@jafar-org/react-components/edit/AsyncSelect'
 */
export default class AsyncSelect extends React.Component {
  static propTypes = {
    value: PropTypes.any,
    state: PropTypes.shape({
      items: PropTypes.array,
      placeholder: PropTypes.string,
      searchQuery: PropTypes.string,
      isLoading: PropTypes.bool,
    }),
    disabled: PropTypes.bool,
    required: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
    onStateChange: PropTypes.func,
  };

  static defaultProps = {
    value: null, // fixes underline controlled VS uncontrolled issue when value turns to undefined 
    state: {
      items: [],
      placeholder: 'Search',
      searchQuery: '',
      isLoading: false,
    },
    disabled: false,
    required: false,
  };

  render() {
    return (
      <SelectInternal
        value={this.props.value}
        onChange={this.props.onValueChange}
        options={this.props.state.items}
        placeholder={this.props.state.placeholder}
        isDisabled={this.props.disabled}
        isClearable={!this.props.required}
        isLoading={!!this.props.state.isLoading}
        isSearchable={true}
        inputValue={this.props.state.searchQuery}
        onInputChange={this.onSearchChange}
        styles={customStyles}
      />
    );
  }

  onSearchChange = (query) => {
    const newState = Object.assign({}, this.props.state);
    newState.searchQuery = query;
    this.props.onStateChange(newState);
  }
}
