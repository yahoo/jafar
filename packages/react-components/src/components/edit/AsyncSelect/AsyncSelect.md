<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const mockServerItems = [{
  id: '1',
  name: 'Ross Geller'
}, {
  id: '2',
  name: 'Monica Geller'
}, {
  id: '3',
  name: 'Rachel Green'
}, {
  id: '4',
  name: 'Chandler Bing'
}, {
  id: '5',
  name: 'Joey Tribbiani'
}, {
  id: '6',
  name: 'Phoebe Buffay'
}];

const mockService = {
  get: (id) => {
    return new Promise(resolve => {
      const item = mockServerItems.find(x => x.id === id);
      resolve(item);
    });
  },
  search: (query) => {
    return new Promise(resolve => {
      const items = mockServerItems.filter(x => x.name.toLowerCase().indexOf(query.toLowerCase()) > -1);
      resolve(items);
    });
  },
};

const model = {
  id: 'simple',
  fields: {
    bestFriend: {
      label: 'Best Friend',
      path: 'bestFriend',
      component: {
        name: 'AsyncSelect',
        stateChange: { name: 'friendStateChange' },
      },
      formatter: { name: 'friendFormatter' },
      parser: { name: 'friendParser' },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
    bestFriend: '3', // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    AsyncSelect: {
      renderer: AsyncSelect,
        stateChange:  (props) => {    
        // if search filter has changed - return object with cleared items and 
        // change isLoading indication to true
        if (props.prevState && (props.prevState.searchQuery !== props.state.searchQuery)) {
          return { ...props.state, items: [], isLoading: true };
        }

        // else, if isLoading changes to true,
        // begin search and return promise which resolves an object of: { items: [ ... ], isLoading: false }
        if (props.state.isLoading) {
          return new Promise((resolve) => {
            // mock server call
            mockService.search(props.state.searchQuery).then((items) => {
              const viewItems = items.map(x => { return { label: x.name, value: x.id }; });
              resolve({ ...props.state, items: viewItems, isLoading: false });
            });
          });
        }

        // else - no more changes needed
        return undefined;
      },
    },
  },
  conversions: {
    friendFormatter: {
      func: (props) => {
        return new Promise((resolve) => {
          // mock server call
          mockService.get(props.value).then((item) => {
            resolve({ label: item.name, value: item.id });
          });
        });  
      } 
    },
    friendParser: {
      func: (props) => {
        return props.value.value; // (props.value is the item { label, value })
      },
    },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='bestFriend' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
  value: { value: '3', label: 'Rachel Green' },
};

<AsyncSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
onStateChange={(newState) => {
  newState.items = [];
  newState.isLoading = true;
  setState({ state: newState });

  setTimeout(() => {
    const items = [{
      value: '1',
      label: 'Ross Geller'
    }, {
      value: '2',
      label: 'Monica Geller'
    }, {
      value: '3',
      label: 'Rachel Green'
    }, {
      value: '4',
      label: 'Chandler Bing'
    }, {
      value: '5',
      label: 'Joey Tribbiani'
    }, {
      value: '6',
      label: 'Phoebe Buffay'
    }];

    const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
    const newState2 = Object.assign({}, newState);
    newState2.items = filteredItems;
    newState2.isLoading = false;
    setState({ state: newState2 });
  }, 700);
}}
/>
```

<h4>Value as object</h4>

```javascript
initialState = { 
  value: { 
    value: { id: '3', first: 'Rachel', last: 'Green' }, 
    label: 'Rachel Green' 
  },
};

<AsyncSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
onStateChange={(newState) => {
  newState.items = [];
  newState.isLoading = true;
  setState({ state: newState });

  setTimeout(() => {
    const items = [{
      value: { id: '1', first: 'Ross', last: 'Geller' },
      label: 'Ross Geller'
    }, {
      value: { id: '2', first: 'Monica', last: 'Geller' },
      label: 'Monica Geller'
    }, {
      value: { id: '3', first: 'Rachel', last: 'Green' },
      label: 'Rachel Green'
    }, {
      value: { id: '4', first: 'Chandler', last: 'Bing' },
      label: 'Chandler Bing'
    }, {
      value: { id: '5', first: 'Joey', last: 'Tribbiani' },
      label: 'Joey Tribbiani'
    }, {
      value: { id: '6', first: 'Phoebe', last: 'Buffay' },
      label: 'Phoebe Buffay'
    }];

    const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
    const newState2 = Object.assign({}, newState);
    newState2.items = filteredItems;
    newState2.isLoading = false;
    setState({ state: newState2 });
  }, 700);
}}
/>
  ```

<h4>Required (cant unselect)</h4>

```javascript
initialState = { 
  value: { value: '3', label: 'Rachel Green' },
};

<AsyncSelect
value={state.value}
state={state.state}
required={true}
onValueChange={(value) => {
  setState({ value });
}}
onStateChange={(newState) => {
  newState.items = [];
  newState.isLoading = true;
  setState({ state: newState });

  setTimeout(() => {
    const items = [{
      value: '1',
      label: 'Ross Geller'
    }, {
      value: '2',
      label: 'Monica Geller'
    }, {
      value: '3',
      label: 'Rachel Green'
    }, {
      value: '4',
      label: 'Chandler Bing'
    }, {
      value: '5',
      label: 'Joey Tribbiani'
    }, {
      value: '6',
      label: 'Phoebe Buffay'
    }];

    const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
    const newState2 = Object.assign({}, newState);
    newState2.items = filteredItems;
    newState2.isLoading = false;
    setState({ state: newState2 });
  }, 700);
}}
/>
```

<h4>Disabled</h4>

```javascript
initialState = { 
  value: { value: '3', label: 'Rachel Green' },
};

<AsyncSelect
value={state.value}
state={state.state}
disabled={true}
onValueChange={(value) => {
  setState({ value });
}}
onStateChange={(newState) => {
  newState.items = [];
  newState.isLoading = true;
  setState({ state: newState });

  setTimeout(() => {
    const items = [{
      value: '1',
      label: 'Ross Geller'
    }, {
      value: '2',
      label: 'Monica Geller'
    }, {
      value: '3',
      label: 'Rachel Green'
    }, {
      value: '4',
      label: 'Chandler Bing'
    }, {
      value: '5',
      label: 'Joey Tribbiani'
    }, {
      value: '6',
      label: 'Phoebe Buffay'
    }];

    const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
    const newState2 = Object.assign({}, newState);
    newState2.items = filteredItems;
    newState2.isLoading = false;
    setState({ state: newState2 });
  }, 700);
}}
/>
```

<h4>Value undefined</h4>

```javascript
initialState = { 
  value: undefined,
};

<AsyncSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
onStateChange={(newState) => {
  newState.items = [];
  newState.isLoading = true;
  setState({ state: newState });

  setTimeout(() => {
    const items = [{
      value: '1',
      label: 'Ross Geller'
    }, {
      value: '2',
      label: 'Monica Geller'
    }, {
      value: '3',
      label: 'Rachel Green'
    }, {
      value: '4',
      label: 'Chandler Bing'
    }, {
      value: '5',
      label: 'Joey Tribbiani'
    }, {
      value: '6',
      label: 'Phoebe Buffay'
    }];

    const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
    const newState2 = Object.assign({}, newState);
    newState2.items = filteredItems;
    newState2.isLoading = false;
    setState({ state: newState2 });
  }, 700);
}}
/>
```
