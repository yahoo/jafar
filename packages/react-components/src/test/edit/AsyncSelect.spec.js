/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import AsyncSelect from '../../components/edit/AsyncSelect/index';

describe('<AsyncSelect />', () => {
  let component;
  let onValueChangeSpy;
  let onStateChangeSpy;
  let state;
  let value;
  let disabled;
  let required;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
    onStateChangeSpy = jest.fn();
    value = { value: 'ocean', label: 'Ocean' };
    disabled = false;
    required = false;
    state = { 
      options:[
        { value: 'ocean', label: 'Ocean' },
        { value: 'blue', label: 'Blue' },
        { value: 'purple', label: 'Purple' },
      ],
    };
  });

  it('Should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render provided data - disabled', () => {
    disabled = true;
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render provided data - required', () => {
    required = true;
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('onSearchChange - ok', () => {
    component = mount(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    const controller = component.instance();
    const query = 'pu';
    controller.onSearchChange(query);
    const expectedState = state;
    expectedState.searchQuery = query;
    expect(onStateChangeSpy).toHaveBeenCalledWith(expectedState);
  });

  function getComponent(value, state, disabled, required, onValueChange, onStateChange) {
    return (
      <AsyncSelect
        value={value}
        state={state}
        disabled={disabled}
        required={required}
        onValueChange={onValueChange}
        onStateChange={onStateChange}
      />
    );
  }
});
