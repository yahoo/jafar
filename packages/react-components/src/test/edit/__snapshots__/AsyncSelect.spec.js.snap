// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AsyncSelect /> Should render provided data - disabled 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isClearable={true}
  isDisabled={true}
  isLoading={false}
  isSearchable={true}
  onChange={[MockFunction]}
  onInputChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Object {
      "label": "Ocean",
      "value": "ocean",
    }
  }
/>
`;

exports[`<AsyncSelect /> Should render provided data - required 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isClearable={false}
  isDisabled={false}
  isLoading={false}
  isSearchable={true}
  onChange={[MockFunction]}
  onInputChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Object {
      "label": "Ocean",
      "value": "ocean",
    }
  }
/>
`;

exports[`<AsyncSelect /> Should render provided data 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isClearable={true}
  isDisabled={false}
  isLoading={false}
  isSearchable={true}
  onChange={[MockFunction]}
  onInputChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Object {
      "label": "Ocean",
      "value": "ocean",
    }
  }
/>
`;
